File: data-structure-and-algorithm/java/wildcard-matching.java
/**
 * Wildcard Matching 
 * 
 * Implement wildcard pattern matching with support for '?' and '*'.
 * 
 * '?' Matches any single character.
 * '*' Matches any sequence of characters (including the empty sequence).
 * 
 * The matching should cover the entire input string (not partial).
 * 
 * The function prototype should be:
 * bool isMatch(const char *s, const char *p)
 * 
 * Some examples:
 * isMatch("aa","a") → false
 * isMatch("aa","aa") → true
 * isMatch("aaa","aa") → false
 * isMatch("aa", "*") → true
 * isMatch("aa", "a*") → true
 * isMatch("ab", "?*") → true
 * isMatch("aab", "c*a*b") → false
 */

public class Solution {
    public boolean isMatch(String s, String p) {
        if (s == null || p == null) {
            throw new NullPointerException();
        }

        int sIndex = 0, pIndex = 0;
        int sStarIndex = 0, pStarIndex = 0;

        while (sIndex < s.length()) {
            if (s.charAt(sIndex) == p.charAt(pIndex) || p.charAt(pIndex) == '?') {
                ++sIndex;
                ++pIndex;
            } else if (p.charAt(pIndex) == '*') {
                pStarIndex = pIndex++;
                sStarIndex = sIndex;
            } else if (pStarIndex > 0) {
                pIndex = pStarIndex + 1;
                sIndex = ++sStarIndex;
            } else {
                return false;
            }
        }
        
        while (p.charAt(pIndex) == '*') {
            ++pIndex;
        }
        return pIndex == p.length();
    }
}

File: data-structure-and-algorithm/typescript/leetcode/src/wildcard-matching.ts
/*
44. Wildcard Matching

Given an input string (s) and a pattern (p), implement wildcard pattern matching with support for '?' and '*' where:

'?' Matches any single character.
'*' Matches any sequence of characters (including the empty sequence).
The matching should cover the entire input string (not partial).

 

Example 1:

Input: s = "aa", p = "a"
Output: false
Explanation: "a" does not match the entire string "aa".
Example 2:

Input: s = "aa", p = "*"
Output: true
Explanation: '*' matches any sequence.
Example 3:

Input: s = "cb", p = "?a"
Output: false
Explanation: '?' matches 'c', but the second letter is 'a', which does not match 'b'.
 

Constraints:

0 <= s.length, p.length <= 2000
s contains only lowercase English letters.
p contains only lowercase English letters, '?' or '*'.
*/
function isMatch(s: string, p: string): boolean {
  let i = 0,
    j = 0;
  let iStar = -1,
    jStar = -1;
  let m = s.length,
    n = p.length;
  while (i < m) {
    if (j < n && (s[i] === p[j] || p[j] === '?')) {
      ++i;
      ++j;
    } else if (j < n && p[j] === '*') {
      iStar = i;
      jStar = j++;
    } else if (iStar >= 0) {
      i = ++iStar;
      j = jStar + 1;
    } else {
      return false;
    }
  }
  while (j < n && p[j] === '*') {
    ++j;
  }
  return j === n;
}
